lib/collection/src/shards/local_shard/query.rs

```diff
@@ -30,65 +32,164 @@ impl LocalShard {
+        let scrolls = &vec![];
+        self.recurse_prefetch(
+            &request.merge_plan,
+            &core_results,
+            scrolls,
+            search_runtime_handle,
+            timeout,
+            0, // initial depth
+        )
+        .await
     }
 
     fn recurse_prefetch<'shard, 'query>(
         &'shard self,
         prefetch: &'query MergePlan,
         core_results: &'query Vec<Vec<ScoredPoint>>,
         scrolls: &'query Vec<Vec<ScoredPoint>>,
-    ) -> BoxFuture<'query, CollectionResult<Vec<ScoredPoint>>>
+        search_runtime_handle: &'shard Handle,
+        timeout: Option<Duration>,
+        depth: usize,
+    ) -> BoxFuture<'query, CollectionResult<Vec<Vec<ScoredPoint>>>>
     where
         'shard: 'query,
     {
         async move {
-            let mut sources = Vec::with_capacity(prefetch.sources.len());
+            let mut sources: Vec<Vec<ScoredPoint>> = Vec::with_capacity(prefetch.sources.len());
 
             for source in prefetch.sources.iter() {
-                let vec: Vec<ScoredPoint> = match source {
+                let vec: Vec<Vec<ScoredPoint>> = match source {
                     PrefetchSource::SearchesIdx(idx) => {
-                        core_results.get(*idx).cloned().unwrap_or_default() // TODO(universal-query): don't clone, by using something like a hashmap instead of a vec
+                        // TODO(universal-query): don't clone, by using something like a hashmap instead of a vec
+                        let scored_searches = core_results.get(*idx).cloned().unwrap_or_default();
+                        vec![scored_searches]
                     }
                     PrefetchSource::ScrollsIdx(idx) => {
-                        scrolls.get(*idx).cloned().unwrap_or_default() // TODO(universal-query): don't clone, by using something like a hashmap instead of a vec
+                        // TODO(universal-query): don't clone, by using something like a hashmap instead of a vec
+                        let scrolled = scrolls.get(*idx).cloned().unwrap_or_default();
+                        vec![scrolled]
                     }
                     PrefetchSource::Prefetch(prefetch) => {
-                        self.recurse_prefetch(prefetch, core_results, scrolls)
-                            .await?
+                        self.recurse_prefetch(
+                            prefetch,
+                            core_results,
+                            scrolls,
+                            search_runtime_handle,
+                            timeout,
+                            depth + 1,
+                        )
+                        .await?
                     }
                 };
-                sources.push(vec);
+                sources.extend(vec);
             }
 
-            self.merge_prefetches(sources, &prefetch.merge).await
+            if depth == 0 && prefetch.merge.rescore == Some(ScoringQuery::Fusion(Fusion::Rrf)) {
```

Comment:
TBH I thought it was going to be handled at `do_planned_query`, but this way there is less repetition. I like this.